lib/ddtrace/contrib/excon/middleware.rb

```diff
@@ -0,0 +1,140 @@
+        # With options given as defaults.
+        def self.with(options = {})
+          Class.new(self) do
+            @options = options
+
+            # rubocop:disable Style/TrivialAccessors
+            def self.options
+              @options
+            end
+
+            def initialize(stack)
+              super(stack, self.class.options)
+            end
+          end
+        end
+
+        # Returns a copy of the default stack with the trace middleware injected
+        def self.around_default_stack
+          ::Excon.defaults[:middlewares].dup.tap do |default_stack|
+            # If the default stack contains a version of the trace middleware already...
+            existing_trace_middleware = default_stack.find { |m| m <= Middleware }
+            default_stack.delete(existing_trace_middleware) if existing_trace_middleware
+
+            # Inject after the ResponseParser middleware
+            response_middleware_index = default_stack.index(::Excon::Middleware::ResponseParser).to_i
+            default_stack.insert(response_middleware_index + 1, self)
+          end
+        end
+
+        private
+
+        def tracer
+          @options[:tracer]
+        end
+
+        def distributed_tracing?
+          @options[:distributed_tracing] == true && tracer.enabled
+        end
+
+        def error_handler
+          @options[:error_handler] || DEFAULT_ERROR_HANDLER
+        end
+
+        def split_by_domain?
+          @options[:split_by_domain] == true
+        end
+
+        def annotate!(span, datum)
+          span.resource = datum[:method].to_s.upcase
+          span.service = service_name(datum)
+          span.span_type = Ext::HTTP::TYPE
+          span.set_tag(Ext::HTTP::URL, datum[:path])
+          span.set_tag(Ext::HTTP::METHOD, datum[:method].to_s.upcase)
+          span.set_tag(Ext::NET::TARGET_HOST, datum[:host])
+          span.set_tag(Ext::NET::TARGET_PORT, datum[:port].to_s)
+        end
+
+        def handle_response(datum)
+          if datum.key?(:datadog_span)
+            datum[:datadog_span].tap do |span|
+              return span if span.finished?
+
+              if datum.key?(:response)
+                response = datum[:response]
+                if error_handler.call(response)
+                  span.set_error(["Error #{response[:status]}", response[:body]])
+                end
+                span.set_tag(Ext::HTTP::STATUS_CODE, response[:status])
+              end
+              span.set_error(datum[:error]) if datum.key?(:error)
+              span.finish
+              datum.delete(:datadog_span)
+            end
+          end
+        rescue StandardError => e
+          Datadog::Tracer.log.debug(e.message)
+        end
+
+        def propagate!(span, datum)
+          Datadog::HTTPPropagator.inject!(span.context, datum[:headers])
+        end
+
+        def service_name(datum)
+          # TODO: Change this to implement more sensible multiplexing
```

Comment:
What is this `TODO` about?

Reply:
This `split_by_domain?` function is generating new services for each host it connects to, named after the host. That seems like it could make a mess of your APM services page double quick. This comment was more to say "should we be splitting the resource by domain instead?"

Reply:
What is expected (and it's how it works for Python) is that if you enable that flag (disabled by default), you should see a new service for each host you're calling (so the name of the service is just the name of the host).

Splitting the resource by domain works too, but it doesn't show the time spent in different services in the sublayer graphs.